File: src/tools/knowledge-base/types/monitor.ts
/**
 * 知识库监控消息类型
 */
export type KbMessageType = "RAG" | "Index" | "Chain" | "System";

/**
 * 监控消息级别
 */
export type KbMonitorLevel = "info" | "warn" | "error" | "success";

/**
 * 步骤状态
 */
export type KbStepStatus = "pending" | "running" | "completed" | "failed";

/**
 * 基础监控消息接口
 */
interface KbMonitorMessageBase {
  /** 唯一标识符 */
  id: string;
  /** 消息级别 */
  level: KbMonitorLevel;
  /** 时间戳 (ms) */
  timestamp: number;
  /** 标题 */
  title: string;
  /** 简要总结 */
  summary: string;
  /** 细分模块，如 "VectorEngine", "FileScanner" */
  module: string;
}

/**
 * RAG 检索追踪数据结构
 */
export interface RagPayload {
  /** 检索步骤详情 */
  steps: Array<{
    /** 步骤名称 */
    name: string;
    /** 状态 */
    status: KbStepStatus;
    /** 耗时 (ms) */
    duration: number;
    /** 详细信息 */
    details?: string;
  }>;
  /** 检索结果片段 */
  results?: Array<{
    /** 文档片段 ID */
    id: string;
    /** 相似度分数 */
    score: number;
    /** 文本内容 */
    content: string;
    /** 来源文件/路径 */
    source?: string;
    /** 额外元数据 */
    metadata?: Record<string, any>;
  }>;
  /** 统计信息 */
  stats: {
    /** 总耗时 (ms) */
    duration: number;
    /** 消耗 Token 数 */
    tokenCount?: number;
    /** 命中的片段数 */
    hitCount?: number;
    /** 召回的原始片段数 */
    recallCount?: number;
  };
  /** 检索上下文元数据 */
  metadata?: {
    /** 查询语句 */
    query: string;
    /** 使用的模型 ID */
    modelId: string;
    /** 使用的引擎 ID */
    engineId: string;
    /** 检索的知识库 ID 列表 */
    kbIds: string[];
  };
}

/**
 * 索引生命周期追踪数据结构
 */
export interface IndexPayload {
  /** 索引步骤详情 */
  steps: Array<{
    /** 步骤名称 */
    name: string;
    /** 状态 */
    status: KbStepStatus;
    /** 耗时 (ms) */
    duration: number;
    /** 详细信息 */
    details?: string;
  }>;
  /** 索引统计信息 */
  stats: {
    /** 总文件数 */
    totalFiles: number;
    /** 已处理文件数 */
    processedFiles: number;
    /** 总切片数 */
    totalChunks: number;
    /** 已向量化切片数 */
    vectorizedChunks: number;
    /** 总耗时 (ms) */
    duration: number;
  };
  /** 索引任务元数据 */
  metadata?: {
    /** 知识库 ID */
    kbId: string;
    /** 使用的模型 ID */
    modelId: string;
    /** 包含的文件模式 */
    filePatterns: string[];
  };
}

/**
 * 链式处理追踪数据结构
 */
export interface ChainPayload {
  /** 链式步骤详情 */
  steps: Array<{
    /** 步骤名称 */
    name: string;
    /** 状态 */
    status: KbStepStatus;
    /** 耗时 (ms) */
    duration: number;
    /** 详细信息 */
    details?: string;
  }>;
  /** 链式任务元数据 */
  metadata?: {
    /** 链类型 */
    chainType: string;
    /** 参数详情 */
    parameters: Record<string, any>;
  };
}

/**
 * 系统级消息数据结构
 */
export interface SystemPayload {
  /** 系统统计指标 */
  stats?: Record<string, number>;
  /** 系统元数据 */
  metadata?: Record<string, any>;
}

/**
 * 核心监控消息类型 (Tagged Union)
 */
export type KbMonitorMessage =
  | (KbMonitorMessageBase & { type: "RAG"; payload: RagPayload })
  | (KbMonitorMessageBase & { type: "Index"; payload: IndexPayload })
  | (KbMonitorMessageBase & { type: "Chain"; payload: ChainPayload })
  | (KbMonitorMessageBase & { type: "System"; payload: SystemPayload });

File: src-tauri/src/knowledge/monitor.rs
use chrono::Utc;
use serde::Serialize;
use std::collections::HashMap;
use tauri::{AppHandle, Emitter};
use uuid::Uuid;

/// 监控消息级别
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum KbMonitorLevel {
    Info,
    #[allow(dead_code)]
    Warn,
    #[allow(dead_code)]
    Error,
    Success,
    #[allow(dead_code)]
    Debug,
}

/// 步骤状态
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum KbStepStatus {
    #[allow(dead_code)]
    Pending,
    Running,
    Completed,
    #[allow(dead_code)]
    Failed,
}

/// 监控步骤详情
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct KbMonitorStep {
    pub name: String,
    pub status: KbStepStatus,
    pub duration: u64,
    pub details: Option<String>,
}

/// RAG 检索追踪数据结构
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RagPayload {
    pub steps: Vec<KbMonitorStep>,
    pub results: Option<Vec<RagResult>>,
    pub stats: RagStats,
    pub metadata: Option<RagMetadata>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RagResult {
    pub id: String,
    pub score: f32,
    pub content: String,
    pub source: Option<String>,
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RagStats {
    pub duration: u64,
    pub token_count: Option<u32>,
    pub hit_count: Option<u32>,
    pub recall_count: Option<u32>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RagMetadata {
    pub query: String,
    pub model_id: String,
    pub engine_id: String,
    pub kb_ids: Vec<String>,
}

/// 索引生命周期追踪数据结构
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IndexPayload {
    pub steps: Vec<KbMonitorStep>,
    pub stats: IndexStats,
    pub metadata: Option<IndexMetadata>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IndexStats {
    pub total_files: u32,
    pub processed_files: u32,
    pub total_chunks: u32,
    pub vectorized_chunks: u32,
    pub duration: u64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IndexMetadata {
    pub kb_id: String,
    pub model_id: String,
    pub file_patterns: Vec<String>,
}

/// 链式处理追踪数据结构
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChainPayload {
    pub steps: Vec<KbMonitorStep>,
    pub metadata: Option<ChainMetadata>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChainMetadata {
    pub chain_type: String,
    pub parameters: HashMap<String, serde_json::Value>,
}

/// 系统级消息数据结构
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SystemPayload {
    pub stats: Option<HashMap<String, f64>>,
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

/// 核心监控事件类型 (Tagged Union)
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", content = "payload")]
pub enum KbMonitorEvent {
    #[allow(clippy::upper_case_acronyms)]
    RAG(RagPayload),
    Index(IndexPayload),
    #[allow(dead_code)]
    Chain(ChainPayload),
    System(SystemPayload),
}

/// 监控消息结构
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct KbMonitorMessage {
    pub id: String,
    pub level: KbMonitorLevel,
    pub timestamp: i64,
    pub title: String,
    pub summary: String,
    pub module: String,
    #[serde(flatten)]
    pub event: KbMonitorEvent,
}

/// 发送监控事件到前端
///
/// 事件名为 "kb-monitor"
pub fn emit_monitor_event(
    app: &AppHandle,
    event: KbMonitorEvent,
    level: KbMonitorLevel,
    title: &str,
    summary: &str,
    module: &str,
) -> Result<(), String> {
    let message = KbMonitorMessage {
        id: Uuid::new_v4().to_string(),
        level,
        timestamp: Utc::now().timestamp_millis(),
        title: title.to_string(),
        summary: summary.to_string(),
        module: module.to_string(),
        event,
    };

    app.emit("kb-monitor", &message)
        .map_err(|e| format!("Failed to emit monitor event: {}", e))
}

/// 发送心跳包
#[tauri::command]
pub async fn kb_monitor_heartbeat(app: AppHandle) -> Result<(), String> {
    emit_monitor_event(
        &app,
        KbMonitorEvent::System(SystemPayload {
            stats: None,
            metadata: Some(
                [("heartbeat".to_string(), serde_json::Value::Bool(true))]
                    .into_iter()
                    .collect(),
            ),
        }),
        KbMonitorLevel::Info,
        "心跳消息",
        "监控系统运行中",
        "System",
    )
}
